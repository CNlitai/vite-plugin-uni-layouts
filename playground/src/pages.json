{
  "pages": [ // pages数组中第一项表示应用启动页，参考：https://uniapp.dcloud.io/collocation/pages
    {
      "path": "pages/index/index",
      "style": {
        "navigationBarTitleText": "default"
      }
    },
    {
      "path": "pages/index/disabled",
      "style": {
        "navigationBarTitleText": "disabled layout"
      },
      "layout": false
    },
    {
      "path": "pages/index/uni-layout",
      "layout": false,
      "style": {
        "navigationBarTitleText": "use uni-layout component"
      }
    }
  ],
  "globalStyle": {
    "navigationBarTextStyle": "black",
    "navigationBarTitleText": "uni-app",
    "navigationBarBackgroundColor": "#F8F8F8",
    "backgroundColor": "#F8F8F8"
  },
  "subPackages": [
    {
      "root": "pages-sub",
      "pages": [
        {
          "path": "test/test",
          "type": "page",
          "layout": "default"
        }
      ]
    }
  ]
}
