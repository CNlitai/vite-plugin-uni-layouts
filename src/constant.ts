export const virtualModuleId = 'virtual:uni-layouts'
