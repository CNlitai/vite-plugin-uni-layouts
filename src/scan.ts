import { basename, dirname, extname, join, relative, resolve } from 'node:path'
import process from 'node:process'
import fg from 'fast-glob'
import { camelCase, kebabCase, pascalCase, splitByCase } from 'scule'
import { normalizePath } from 'vite'
import type { Layout } from './types'

export function scanLayouts(dir = 'src/layouts', cwd = process.cwd()) {
  dir = resolve(cwd, dir)
  const files = fg.sync('**/*.vue', {
    onlyFiles: true,
    ignore: ['node_modules', '.git', '**/__*__/*'],
    cwd: dir,
  })
  files.sort()

  const layouts: Layout[] = []

  for (const file of files) {
    const filePath = normalizePath(join(dir, file))
    const dirNameParts = splitByCase(
      normalizePath(relative(dir, dirname(filePath))),
    )
    let fileName = basename(filePath, extname(filePath))
    if (fileName.toLowerCase() === 'index')
      fileName = basename(dirname(filePath))

    const fileNameParts = splitByCase(fileName)
    const componentNameParts: string[] = []

    while (
      dirNameParts.length
      && (dirNameParts[0] || '').toLowerCase()
        !== (fileNameParts[0] || '').toLowerCase()
    )
      componentNameParts.push(dirNameParts.shift()!)

    const pascalName
      = pascalCase(componentNameParts) + pascalCase(fileNameParts)
    const camelName = camelCase(pascalName)
    const kebabName = kebabCase(pascalName)
    layouts.push({
      name: camelName,
      path: filePath,
      pascalName,
      kebabName,
    })
  }
  return layouts
}
