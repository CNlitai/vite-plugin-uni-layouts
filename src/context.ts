import { resolve } from 'node:path'
import type { FSWatcher, ResolvedConfig, ViteDevServer } from 'vite'
import { normalizePath } from 'vite'
import MagicString from 'magic-string'
import type { ElementNode } from '@vue/compiler-dom'
import { parse } from '@vue/compiler-dom'
import { kebabCase } from 'scule'
import { walk } from 'estree-walker'
import { scanLayouts } from './scan'
import type { Layout, Page, ResolvedOptions } from './types'
import { getTarget, loadPagesJson } from './utils'

export class Context {
  config!: ResolvedConfig
  options: ResolvedOptions
  pages: Page[]
  layouts: Layout[]
  private _server?: ViteDevServer
  parse?: (input: string, options?: any) => any
  constructor(options: ResolvedOptions) {
    this.options = options
    this.pages = []
    this.layouts = scanLayouts(options.layoutDir, options.cwd)
  }

  setupViteServer(server: ViteDevServer) {
    if (this._server === server)
      return

    this._server = server
    this.setupWatcher(server.watcher)
  }

  async setupWatcher(watcher: FSWatcher) {
    watcher.on('change', async (path) => {
      if (
        normalizePath(path)
        === normalizePath(resolve(this.options.cwd, 'src/pages.json'))
      )
        this.pages = loadPagesJson('src/pages.json', this.options.cwd)
      // TODO: auto reload
    })
  }

  transform(code: string, path: string) {
    if (!this.pages?.length)
      this.pages = loadPagesJson('src/pages.json', this.options.cwd)

    const page = getTarget(
      path,
      this.pages,
      this.options.layout,
      this.config.root,
    )
    if (!page)
      return
    if (!this.layouts.length)
      return
    let layoutName: string | undefined | false = page.layout
    let layout: Layout | undefined

    if (typeof layoutName === 'boolean' && layoutName)
      layoutName = 'default'

    if (typeof layoutName === 'string') {
      if (!layoutName)
        return
      layout = this.layouts.find(l => l.name === (layoutName as string))
      if (!layout)
        return
    }

    const ast = parse(code)
    const ms = new MagicString(code)
    let sourceWithoutRoot = ''
    let props: string[] = []
    let dynamicLayout = ''
    const rootTemplate = ast.children.find(
      node => node.type === 1 && node.tag === 'template',
    ) as ElementNode
    const rootScript = ast.children.find(
      node => node.type === 1 && node.tag === 'script',
    ) as ElementNode
    if (!rootTemplate)
      return

    if (rootScript) {
      const firstContent = rootScript.children?.[0]
      const code = firstContent?.type === 2 && firstContent.content
      if (code && this.parse) {
        const ast = this.parse(code)
        walk(ast, {
          enter(node) {
            if (node.type === 'VariableDeclarator') {
              const hasUniLayoutVar
                = node.id.type === 'Identifier' && node.id.name === 'uniLayout'
              const isRef
                = node.init?.type === 'CallExpression'
                && node.init.callee.type === 'Identifier'
                && node.init.callee.name === 'ref'
              if (hasUniLayoutVar && isRef)
                props.push('ref="uniLayout"')
            }
          },
        })
      }
    }

    const isDisabledLayout = typeof layoutName === 'boolean' && !layoutName
    if (isDisabledLayout) {
      const uniLayoutComponent = rootTemplate.children.find(
        node => node.type === 1 && kebabCase(node.tag) === 'uni-layout',
      ) as ElementNode

      if (uniLayoutComponent) {
        props = uniLayoutComponent.props.map(v => v.loc.source)
        for (const prop of uniLayoutComponent.props) {
          if (
            prop.name === 'bind'
            && prop.type === 7
            && prop?.exp?.type === 4
            && prop.arg?.type === 4
            && prop.arg?.content === 'name'
          )
            dynamicLayout = prop.exp.content

          if (prop.name === 'name' && prop.type === 6) {
            layoutName = prop.value?.content
            // not set layout
            if (!layoutName)
              return
            layout = this.layouts.find(
              l => l.name === (layoutName as string),
            )
            if (!layout)
              return
          }
        }
        sourceWithoutRoot += uniLayoutComponent.children
          .map(v => v.loc.source)
          .join('\n')
      }
      else {
        return
      }
    }
    else {
      sourceWithoutRoot += rootTemplate.children
        .map(v => v.loc.source)
        .join('')
    }
    ms.replace(rootTemplate.loc.source, '')
    if (dynamicLayout) {
      ms.prepend(`<template>
  <component ${props.join(' ')} :is='\`layout-\${${dynamicLayout}}-uni\`'>
    ${sourceWithoutRoot}
  </component>
</template>
`)
    }
    else {
      ms.prepend(`<template>
  <layout-${layout?.kebabName}-uni ${props.join(' ')}>
    ${sourceWithoutRoot}
  </layout-${layout?.kebabName}-uni>
</template>
`)
    }
    const map = ms.generateMap({
      source: path,
      file: `${path}.map`,
      includeContent: true,
    })
    return {
      code: ms.toString(),
      map,
    }
  }

  async virtualModule() {
    const imports: string[] = []
    const components: string[] = []
    const _exports = this.layouts.map((v) => {
      imports.push(
        `import Layout_${v.pascalName}_Uni from "${normalizePath(v.path)}"`,
      )
      components.push(
        `app.component("layout-${v.kebabName}-uni", Layout_${v.pascalName}_Uni)`,
      )
      return `Layout_${v.pascalName}_Uni,`
    })
    return `${imports.join('\n')}
export const layouts = {
  ${_exports.join('\n')}
}
export default {
  install(app) {
    ${components.join('\n')}
  }
}`
  }

  async importLayoutComponents(code: string, id: string) {
    const ms = new MagicString(code)
    const imports: string[] = []
    const components: string[] = []
    for (const v of this.layouts) {
      imports.push(
        `import Layout_${v.pascalName}_Uni from "${normalizePath(v.path)}"`,
      )
      components.push(
        `app.component("layout-${v.kebabName}-uni", Layout_${v.pascalName}_Uni);\n`,
      )
    }
    ms.append(imports.join('\n'))
    ms.replace(
      /(createApp[\s\S]*?)(return\s{\s*app)/,
      `$1${components.join('')}$2`,
    )
    const map = ms.generateMap({
      source: id,
      file: `${id}.map`,
      includeContent: true,
    })
    code = ms.toString()
    return {
      code,
      map,
    }
  }
}
