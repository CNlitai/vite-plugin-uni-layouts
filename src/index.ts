import type { Plugin } from 'vite'
import { createFilter } from 'vite'
import { virtualModuleId } from './constant'
import { Context } from './context'
import type { UserOptions } from './types'
import { resolveOptions } from './utils'

export function VitePluginUniLayouts(userOptions: UserOptions = {}): Plugin {
  const options = resolveOptions(userOptions)
  const ctx = new Context(options)
  return {
    name: 'vite-plugin-uni-layouts',
    enforce: 'pre',
    configResolved(config) {
      ctx.config = config
    },
    configureServer(server) {
      ctx.setupViteServer(server)
    },
    resolveId(id) {
      if (id === virtualModuleId)
        return id
    },
    load(id) {
      if (id === virtualModuleId)
        return ctx.virtualModule()
    },
    transform(code, id) {
      ctx.parse = this.parse
      const filter = createFilter('src/main.(ts|js)')
      if (filter(id))
        return ctx.importLayoutComponents(code, id)

      return ctx.transform(code, id)
    },
  }
}

export default VitePluginUniLayouts
