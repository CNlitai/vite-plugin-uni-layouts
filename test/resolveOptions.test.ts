import path, { } from 'node:path'
import { expect, it } from 'vitest'
import { resolveOptions } from '../src/utils'

it('resolve user options', () => {
  const options = resolveOptions({
    layout: 'home',
    layoutDir: 'src/layout',
  })
  options.cwd = path.basename(options.cwd)
  expect(options).toMatchSnapshot('userOptions')
})

it('resolve default options', () => {
  const options = resolveOptions()
  options.cwd = path.basename(options.cwd)
  expect(options).toMatchSnapshot('default')
})
