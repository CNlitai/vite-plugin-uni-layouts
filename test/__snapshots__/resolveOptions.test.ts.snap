// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolve default options > default 1`] = `
{
  "cwd": "vite-plugin-uni-layouts",
  "layout": "default",
  "layoutDir": "src/layouts",
}
`;

exports[`resolve user options > userOptions 1`] = `
{
  "cwd": "vite-plugin-uni-layouts",
  "layout": "home",
  "layoutDir": "src/layout",
}
`;
